src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Infer.java

```diff
@@ -651,6 +652,81 @@ public Type instantiateFunctionalInterface(DiagnosticPosition pos, Type funcInte
             return owntype;
         }
     }
+
+    public Type instantiatePatternType(Type expressionType, TypeSymbol patternTypeSymbol) {
+        if (expressionType.tsym == patternTypeSymbol)
+            return expressionType;
+
+        //step 1:
+        Type expressionTypeCaptured = types.capture(expressionType);
+        List<Type> params = patternTypeSymbol.type.allparams();
+        List<Type> capturedWildcards = List.nil();
+        //add synthetic captured ivars
+        for (Type ta : expressionTypeCaptured.getTypeArguments()) {
+            if (ta.hasTag(TYPEVAR) && ((TypeVar)ta).isCaptured()) {
+                params = params.prepend((TypeVar)ta);
+                capturedWildcards = capturedWildcards.prepend(ta);
+            }
+        }
+        InferenceContext c = new InferenceContext(this, params);
+        Type patternType = c.asUndetVar(patternTypeSymbol.type);
+        Type exprType = c.asUndetVar(expressionTypeCaptured);
+
+        capturedWildcards.forEach(s -> ((UndetVar) c.asUndetVar(s)).setNormal());
+
+        //step 2:
+        Set<Symbol> patternTypeSuperTypes = new HashSet<>();
+        types.closure(patternTypeSymbol.type).stream().map(s -> s.tsym).forEach(patternTypeSuperTypes::add);
+        Set<Symbol> expressionTypeSuperTypes = new HashSet<>();
+        types.closure(expressionType).stream().map(s -> s.tsym).forEach(expressionTypeSuperTypes::add);
+        patternTypeSuperTypes.retainAll(expressionTypeSuperTypes);
+
+        for (Symbol common : patternTypeSuperTypes) {
+            Type fromPatternType = types.asSuper(patternType, common);
+            Type fromExprType = types.asSuper(exprType, common);
+            if (!types.isSameType(fromPatternType, fromExprType) && !fromExprType.isRaw()) {
+                return null;
+            }
+        }
+
+        List<Type> varsToSolve = params.map(s -> c.asUndetVar(s));
+
+        try {
+            doIncorporation(c, types.noWarnings);
+
+            while (c.solveBasic(varsToSolve, EnumSet.of(InferenceStep.EQ)).nonEmpty()) {
+                doIncorporation(c, types.noWarnings);
+            }
+        } catch (Infer.InferenceException ex) {
+            return null;
+        }
+
+        //step 3:
+        ListBuffer<Type> freshVars = new ListBuffer<>();
+
+        for (Type param : patternType.allparams()) {
+            UndetVar undet = (UndetVar) param;
+            List<Type> bounds = InferenceStep.EQ.filterBounds(undet, c);
+            if (bounds.nonEmpty()) {
+                undet.setInst(bounds.head);
+            } else {
+                List<Type> upperBounds = undet.getBounds(InferenceBound.UPPER);
+                Type bound = upperBounds.isEmpty() ? syms.objectType : types.glb(upperBounds);
```

Comment:
I'm not sure this is 100% correct. E.g. what if an under variable has an upper bound that refers to itself - e.g. `T <: Foo<T>`. In this case we need to step - first create the new tvars, then replace new tvars in the bound of the old ones. We do this in `Infer::instantiateAsUninferredVars` - perhaps some code from there should be borrowed?